runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1396,163 +1392,36 @@ private WorkItemCommitRequest buildWorkItemTruncationRequest(
-        ComputationState computationState = entry.getKey();
-        for (Windmill.WorkItemCommitRequest workRequest : entry.getValue().getRequestsList()) {
-          computationState.completeWorkAndScheduleNextWorkForKey(
-              ShardedKey.create(workRequest.getKey(), workRequest.getShardingKey()),
-              WorkId.builder()
-                  .setCacheToken(workRequest.getCacheToken())
-                  .setWorkToken(workRequest.getWorkToken())
-                  .build());
-        }
-      }
-    }
-  }
-
-  // Adds the commit to the commitStream if it fits, returning true iff it is consumed.
-  private boolean addCommitToStream(Commit commit, CommitWorkStream commitStream) {
-    Preconditions.checkNotNull(commit);
-    final ComputationState state = commit.computationState();
-    final Windmill.WorkItemCommitRequest request = commit.request();
-    // Drop commits for failed work. Such commits will be dropped by Windmill anyway.
-    if (commit.work().isFailed()) {
-      readerCache.invalidateReader(
-          WindmillComputationKey.create(
-              state.getComputationId(), request.getKey(), request.getShardingKey()));
-      stateCache
-          .forComputation(state.getComputationId())
-          .invalidate(request.getKey(), request.getShardingKey());
-      state.completeWorkAndScheduleNextWorkForKey(
-          ShardedKey.create(request.getKey(), request.getShardingKey()),
-          WorkId.builder()
-              .setWorkToken(request.getWorkToken())
-              .setCacheToken(request.getCacheToken())
-              .build());
-      return true;
-    }
-
-    final int size = commit.getSize();
-    commit.work().setState(Work.State.COMMITTING);
-    activeCommitBytes.addAndGet(size);
-    if (commitStream.commitWorkItem(
-        state.getComputationId(),
-        request,
-        (Windmill.CommitStatus status) -> {
-          if (status != Windmill.CommitStatus.OK) {
-            readerCache.invalidateReader(
-                WindmillComputationKey.create(
-                    state.getComputationId(), request.getKey(), request.getShardingKey()));
-            stateCache
-                .forComputation(state.getComputationId())
-                .invalidate(request.getKey(), request.getShardingKey());
-          }
-          activeCommitBytes.addAndGet(-size);
-          state.completeWorkAndScheduleNextWorkForKey(
-              ShardedKey.create(request.getKey(), request.getShardingKey()),
-              WorkId.builder()
-                  .setCacheToken(request.getCacheToken())
-                  .setWorkToken(request.getWorkToken())
-                  .build());
-        })) {
-      return true;
-    } else {
-      // Back out the stats changes since the commit wasn't consumed.
-      commit.work().setState(Work.State.COMMIT_QUEUED);
-      activeCommitBytes.addAndGet(-size);
-      return false;
-    }
+  private void onStreamingCommitFailed(Commit commit) {
```

Comment:
can we get rid of the separate failed handler? Can we just create a CompletedCommit with appropriate error status and use single onCompletedCommit handler?

Reply:
done